dali/operators/reader/numpy_reader_gpu_op_impl.cu

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/slice/slice_gpu.cuh"
+#include "dali/kernels/slice/slice_kernel_utils.h"
+#include "dali/kernels/transpose/transpose_gpu.h"
+#include "dali/operators/reader/numpy_reader_gpu_op.h"
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+template <typename T, int Dims>
+void NumpyReaderGPU::RunImplTyped(DeviceWorkspace &ws) {
+  auto &output = ws.OutputRef<GPUBackend>(0);
+  const auto &out_sh = output.shape();
+  int ndim = out_sh.sample_dim();
+  int nsamples = out_sh.num_samples();
+  bool need_slice = !rois_.empty();
+  auto out_view = view<T>(output);
+  auto curr_batch = GetCurrBatchView<T, Dims>();
+  auto dtype = TypeTable::GetTypeInfoFromStatic<T>();
+
+  // Permuted dims, to use for the transposition
+  std::array<int, Dims> perm;
+  for (int d = 0; d < Dims; d++)
+    perm[d] = Dims - 1 - d;
+
+  int nsamples_copy = 0, nsamples_slice = 0, nsamples_transpose = 0, nsamples_slice_transpose = 0;
+  for (int i = 0; i < nsamples; i++) {
```

Comment:
Now what I think about it maybe it should just have 3 vectors, with indices:
```
for (int i = 0; i < nsamples; i++) {

    if (need_slice_[i] && need_transpose_[i])
      nsamples_slice_transpose.push_back(i);
    else if (need_slice_[i])
      nsamples_slice.push_back(i);
    else if (need_transpose_[i])
      nsamples_transpose.push_back(i);
    else
      nsamples_copy.push_back(i);
  }
```
So you don't have to iterate over and over all samples.

Reply:
Just a suggestion.

Reply:
As discussed offline, the resulting logic would be more complicated than it is now. So, I will keep it as is.